webapp/dojo_news/src/App.spec.js

```diff
@@ -0,0 +1,221 @@
+        expect(wrapper.findAll(".newslistitem").length).toBe(1);
+
+        wrapper.vm.removeNewsListItem(0);
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(0);
+
+        wrapper.vm.createNewsListItem("Test");
+        wrapper.vm.createNewsListItem("Test2");
+
+        await Vue.nextTick();
+
+        expect(wrapper.find("#newslist").html()).not.toBe('<div id="newslist"></div>');
+        expect(wrapper.findAll(".newslistitem").length).toBe(2);
+    });
+});
+
+describe("List Empty Message", () => {
+    const listEmptyMessage = "Oh noes, the list is empty!";
+
+    it('displays a message on an initially empty list of NewsItems', () => {
+        const wrapperWithEmptyList = shallowMount(App, {
+            propsData: {
+                listEmptyMessage: listEmptyMessage,
+                initialNewsListItems: []
+            }
+        });
+        expect(wrapperWithEmptyList.text()).toContain(listEmptyMessage);
+    });
+
+    it('displays no message on an initially non-empty list of NewsItems', () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                listEmptyMessage: listEmptyMessage,
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 0},
+                    {id: 1, title: "Linux", votes: 0},
+                    {id: 2, title: "Windows", votes: 0}
+                ]
+            }
+        });
+
+        expect(wrapper.text()).not.toContain(listEmptyMessage);
+    });
+
+    it('adds and removes the message on list mutations', async () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                listEmptyMessage: listEmptyMessage,
+                initialNewsListItems: []
+            }
+        });
+
+        expect(wrapper.text()).toContain(listEmptyMessage);
+
+        wrapper.vm.createNewsListItem("Test")
+
+        await Vue.nextTick();
+
+        expect(wrapper.text()).not.toContain(listEmptyMessage);
+
+        wrapper.vm.removeNewsListItem(0);
+
+        await Vue.nextTick();
+
+        expect(wrapper.html()).toContain(listEmptyMessage);
+    });
+});
+
+describe("Reverse Order Button", () => {
+    it('list order is reversed upon Reverse-Button click from initial state _descending_', async () => {
+        const wrapper = shallowMount(App, {
+            propsData: {
+                initialNewsListItems: [
+                    {id: 0, title: "macOS", votes: 4},
+                    {id: 1, title: "Linux", votes: 8},
+                    {id: 2, title: "centOS", votes: 3},
+                    {id: 3, title: "Windows", votes: 0}
+                ],
+                initialDescendingOrder: true
+            }
+        });
+
+        let expectedIdList = [1, 0, 2, 3];
+        // or check HTML
+        let actualIdList = wrapper.vm.sortedNewsListItems.map(x => x.id);
```

Comment:
Here you're testing against the internal property`sortedNewsListItems` and not what's being rendered. This puts you under a high risk of false-negatives.

Reply:
I tried to fix that but it's a pretty ugly solution in my opinion. I am testing against the `wrapper.find("newslist").element.innerHTML`. It feels very wrong because I obtained the "correct" string by printing this exact thing and looking through it for possible errors. And this is still not the final string that actually gets rendered by the browser ... Do you have a better way of doing this in mind?

Reply:
expect(wrapper.findAllComponents(NewsItem).wrappers.map(w => w.text())).toEqual(['Title a (2)', 'Title b (1)'])